### PR TITLE
Fix command ID, fix install of the extension only once

### PR DESCRIPTION
**Pull request by AndrewAdov**

### CODE REVIEW

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs [hunk 1]
```diff
@@ -84,6 +84,8 @@ public void WriteLine(string format, params object[] args)
             {
                 _testOutputHelper.WriteLine(format, args);
             }
+
+            public override object InitializeLifetimeService() => null;
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs:88]: ❔ Why is this one long-lived? It should only be valid for the duration of a single call, so won't the infrastructure keep it alive for the necessary duration?

File: src/Microsoft.VisualStudio.IntegrationTestService.Shared/IntegrationService.cs [hunk 2]
```diff
@@ -50,6 +50,8 @@ public static IntegrationService GetInstanceFromHostProcess(Process hostProcess)
             return (IntegrationService)Activator.GetObject(typeof(IntegrationService), uri);
         }
 
+        public override object InitializeLifetimeService() => null;
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.IntegrationTestService.Shared/IntegrationService.cs:53]: ❔ Won't `IntegrationTestServiceCommands._marshalledService` keep this alive for the correct duration anyway?

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs (hunk 1 above)
Comment (AndrewAdov) [reply src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs:88]: Hm... We have the error "IPC channel failure" (VS2019). I assumed that this is due to the fact that the object is destroyed earlier than it should. Now I can definitely confirm that this is definitely not the case for this reason. In 2019 studio, during testing, communication with the VB/CS compiler constantly disappears, which leads to a drop in the IPC channel.

Comment (AndrewAdov) [reply src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs:88]: Nevertheless, we managed to get around this error on our side, so you can skip the part with the InitializeLifetimeService.

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/Harness/IdeTestAssemblyRunner.cs [hunk 3]
```diff
@@ -179,25 +179,87 @@ protected override async Task<RunSummary> RunTestCollectionAsync(IMessageBus mes
...
+                            {
+                                needRestart = true;
+                            }
+                            catch (System.InvalidOperationException ex)
+                            {
+                                if (ex?.Message == "Failed to set the foreground window.")
+                                {
+                                    needRestart = true;
+                                }
+                            }
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/Harness/IdeTestAssemblyRunner.cs:230]: This is fixed by #81

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/Harness/IdeTestAssemblyRunner.cs [hunk 4]
```diff
@@ -238,26 +301,26 @@ private static Version GetVersion(VisualStudioVersion visualStudioVersion)
         {
             switch (visualStudioVersion)
             {
-            case VisualStudioVersion.VS2012:
-                return new Version(11, 0);
+                case VisualStudioVersion.VS2012:
+                    return new Version(11, 0);
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/Harness/IdeTestAssemblyRunner.cs:305]: This is fixed by #80

File: global.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "5.0.302",
+    "version": "5.0.400",
```
Comment (sharwell) [inline global.json:3]: 📝 SDK updates should be sent as a separate pull request

File: src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs [hunk 6]
```diff
@@ -18,6 +18,11 @@ public static class Installer
         public static void Main(string[] args)
         {
             var rootSuffix = args[0];
+            if (rootSuffix == "NoExp")
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs:21]: 📝 I have no objection to supporting customized root suffix, but I would prefer to see it added as a property to `IdeFactAttribute`.

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs (hunk 1 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs:88]: > communication with the VB/CS compiler constantly disappears

How is this observed?

File: global.json (hunk 5 above)
Comment (sharwell) [reply global.json:3]: I'll submit a PR to update to 6.x

File: src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs (hunk 6 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs:21]: I submitted #82 to provide a place (attribute) to specify the root suffix used for all test cases in an entire class.

**Comment by AndrewAdov** [comment]
This request contains changes for internal use only. I will create the separated request later.

**Comment by sharwell** [comment]
@AndrewAdov I noticed during testing today that the default Xunit behavior is one test collection for each test class, and this behavior resulted in the test framework restarting Visual Studio between the tests in each class. I was able to get the tests to share an instance by adding the following:

```
[assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly, DisableTestParallelization = true)]
```

**Comment by AndrewAdov** [comment]
> @AndrewAdov I noticed during testing today that the default Xunit behavior is one test collection for each test class, and this behavior resulted in the test framework restarting Visual Studio between the tests in each class. I was able to get the tests to share an instance by adding the following:
> 
> ```
> [assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly, DisableTestParallelization = true)]
> ```

Sure. I'm pretty sure the extension shouldn't be reinstalled every time the visual studio is restarted.

**Comment by sharwell** [comment]
Agreed, but haven't made it to that part yet. The CollectionBehavior change makes it less critical that the restart behavior be fixed though.

File: src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs (hunk 6 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.VsixInstaller.Shared/Installer.cs:21]: ➡️ I submitted #85 to start addressing this 🎉

File: src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs (hunk 1 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/OutOfProcess/TestInvoker_OutOfProc.cs:88]: @AndrewAdov I submitted #91 with several changes related to this
